celery/backends/redis.py

```diff
@@ -99,6 +99,7 @@ def on_after_fork(self):
         super().on_after_fork()
 
     def _reconnect_pubsub(self):
+        self._pubsub.close()
```

Comment:
adding this remove the error in master branch/5.2.b2/b3 pypi releases as well right?

Reply:
I think so. But this error is hard to reproduce. I'm not sure how to write a test case for it. It's a race condition. It happens only in the production environment in our case (though pretty often to cause serious annoyance - several times a week, but sometimes only once in a month - I guess because it depends on Redis connectivity issues).

Also, looking at the code, `self._pubsub` should be garbage collected immediately by setting it to None, but that doesn't happen. It is collected only when garbage collector runs a bit later. Probably there's some other reference to `self._pubsub` which I was unable to identify by looking at the code. And I don't have the luxury to debug in our production environment. I was lucky enough to get the gdb traceback once.

I've looked at the backend code and it looks identical to the backend code of Celery 4.4.7. So I'm pretty sure the deadlock would happen in the master branch as well.

But if such guesses are not enough I'll try to think on how to create a bulletproof test case proving the need for a change.

Also, quick search shows that other Celery users also experience random lockups when using Redis backend.

Reply:
as 4.4.7 is EOL now we need a patch on master branch only.

Reply:
I did a patch for master branch only.

In our environment I patch celery 4.4.7 by monkey patching. Still waiting results if this would help (second day it is running fine without lock ups, but it's too early to tell).

This part of the Redis backend code haven't changed since 4.4.7.

Reply:
Just checked a history of the redis backend manually, and I don't see any change between 4.4.7 and master branch that could alleviate this problem. The code was hardly changed since then.